dropshot/src/handler.rs

```diff
@@ -268,17 +268,38 @@ where
     ) -> Result<Response<Body>, HandlerError>;
 }
 
-/// Errors returned by an endpoint handler function.
+/// Errors returned by [`HttpHandlerFunc::handle_request`].
+/// `handle_request` may return an error in the following cases:
 ///
-/// A function implementing an HTTP endpoint must return a `Result` with this
-/// error type. Typically, user code will only interact with this type via its
-/// `From` implementation, which allows it to be constructed from any value
-/// implementing the [`HttpResponseError`] trait. This type is largely an
-/// implementation detail, which is necessary to represent errors which may be
-/// errors returned by the endpoint's handler function *or* generated internally
-/// (i.e. by extractors).
+/// 1. A request extractor fails before calling the user-defined handler
+///    function.
+/// 2. The handler function itself is fallible and returns an error.
+/// 3. The response returned by the handler function fails to be converted into
+///    a `Response<Body>`.
+///
+/// In cases (1) and (3), the error will always be an [`HttpError`], but in case
```

Comment:
Oof! As discussed in DMs, this is kind of awful. In particular it means that sometimes clients will see the custom error type and sometimes not! It's also a mess with regard to OpenAPI. We **could** just say "the error is `oneOf` these types", and that's awful too!

As discussed, we might add conversion methods to `HttpResponseError` for cases 1 and 3 here.